File: cloudflare_agent/example_usage.ts
/**
 * Ejemplo de uso del agente A2E en Cloudflare
 * 
 * Este archivo muestra cómo usar el agente A2E para ejecutar workflows
 */

import { A2EAgent, A2EAgentState, A2EAgentEnv } from './a2e_agent';

/**
 * Ejemplo 1: Configurar y obtener capacidades
 */
export async function example1_GetCapabilities(env: A2EAgentEnv) {
	const agent = new A2EAgent(env, {
		a2eServerUrl: 'http://localhost:8000',
		apiKey: 'your-api-key-here',
		executions: [],
	});

	// Configurar servidor A2E
	await agent.configureA2E('http://localhost:8000', 'your-api-key-here');

	// Obtener capacidades
	const capabilities = await agent.getCapabilities();
	console.log('Available APIs:', Object.keys(capabilities.capabilities?.availableApis || {}));
	console.log('Supported Operations:', capabilities.capabilities?.supportedOperations || []);
}

/**
 * Ejemplo 2: Ejecutar un workflow manual
 */
export async function example2_ExecuteManualWorkflow(env: A2EAgentEnv) {
	const agent = new A2EAgent(env, {
		a2eServerUrl: 'http://localhost:8000',
		apiKey: 'your-api-key-here',
		executions: [],
	});

	await agent.configureA2E('http://localhost:8000', 'your-api-key-here');

	// Workflow simple: hacer una llamada API
	const workflowJsonl = `{"operationUpdate": {"workflowId": "test-workflow", "operations": [
		{"id": "fetch", "operation": {
			"ApiCall": {
				"method": "GET",
				"url": "https://api.example.com/users",
				"headers": {
					"Authorization": {
						"credentialRef": {"id": "api-token"}
					}
				},
				"outputPath": "/workflow/users"
			}
		}}
	]}}
{"beginExecution": {"workflowId": "test-workflow", "root": "fetch"}}`;

	// Validar
	const validation = await agent.validateWorkflow(workflowJsonl);
	console.log('Validation:', validation);

	if (validation.valid) {
		// Ejecutar
		const result = await agent.executeWorkflow(workflowJsonl);
		console.log('Execution result:', result);
	}
}

/**
 * Ejemplo 3: Generar y ejecutar workflow usando LLM
 */
export async function example3_GenerateAndExecute(env: A2EAgentEnv) {
	const agent = new A2EAgent(env, {
		a2eServerUrl: 'http://localhost:8000',
		apiKey: 'your-api-key-here',
		executions: [],
	});

	await agent.configureA2E('http://localhost:8000', 'your-api-key-here');

	// Generar y ejecutar workflow usando LLM
	const description = 'Obtener usuarios activos de la API y filtrar los que tienen más de 100 puntos';
	
	const result = await agent.generateAndExecuteWorkflow(description, true);
	console.log('Generated and executed workflow:', result);
}

/**
 * Ejemplo 4: Buscar conocimiento usando RAG
 */
export async function example4_SearchKnowledge(env: A2EAgentEnv) {
	const agent = new A2EAgent(env, {
		a2eServerUrl: 'http://localhost:8000',
		apiKey: 'your-api-key-here',
		executions: [],
	});

	await agent.configureA2E('http://localhost:8000', 'your-api-key-here');

	// Buscar conocimiento relevante
	const results = await agent.searchKnowledge('cómo obtener usuarios', undefined, undefined, 5);
	console.log('Knowledge search results:', results);
}

/**
 * Ejemplo 5: Buscar consultas SQL
 */
export async function example5_SearchSQLQueries(env: A2EAgentEnv) {
	const agent = new A2EAgent(env, {
		a2eServerUrl: 'http://localhost:8000',
		apiKey: 'your-api-key-here',
		executions: [],
	});

	await agent.configureA2E('http://localhost:8000', 'your-api-key-here');

	// Buscar consultas SQL relevantes
	const results = await agent.searchSQLQueries('obtener usuarios activos', 'users_db', 'analytics', 5);
	console.log('SQL queries found:', results);
}

/**
 * Ejemplo 6: Obtener historial de ejecuciones
 */
export async function example6_GetHistory(env: A2EAgentEnv) {
	const agent = new A2EAgent(env, {
		a2eServerUrl: 'http://localhost:8000',
		apiKey: 'your-api-key-here',
		executions: [],
	});

	await agent.configureA2E('http://localhost:8000', 'your-api-key-here');

	// Obtener historial local
	const localHistory = await agent.getExecutionHistory(10);
	console.log('Local execution history:', localHistory);

	// Obtener historial del servidor
	const serverHistory = await agent.listExecutions(100);
	console.log('Server execution history:', serverHistory);
}

File: google_adk_agent/example_usage.py
"""
Ejemplos de uso del agente A2E con Google ADK
"""

import asyncio
from a2e_agent import create_a2e_agent, A2EAgent


async def example1_basic_usage():
    """Ejemplo 1: Uso básico del agente"""
    print("=" * 60)
    print("Ejemplo 1: Uso Básico")
    print("=" * 60)
    
    # Crear agente
    agent = create_a2e_agent(
        a2e_server_url="http://localhost:8000",
        api_key="your-api-key-here"
    )
    
    # Obtener capacidades
    print("\n1. Obteniendo capacidades...")
    capabilities = agent.get_capabilities()
    print(f"   APIs disponibles: {list(capabilities.get('capabilities', {}).get('availableApis', {}).keys())}")
    print(f"   Operaciones: {capabilities.get('capabilities', {}).get('supportedOperations', [])}")
    
    # Interactuar con el agente
    print("\n2. Consultando al agente...")
    response = await agent.run("¿Qué capacidades tengo disponibles en el servidor A2E?")
    print(f"   Respuesta: {response.content}")


async def example2_execute_workflow():
    """Ejemplo 2: Ejecutar un workflow"""
    print("\n" + "=" * 60)
    print("Ejemplo 2: Ejecutar Workflow")
    print("=" * 60)
    
    agent = create_a2e_agent(
        a2e_server_url="http://localhost:8000",
        api_key="your-api-key-here"
    )
    
    # Workflow simple: esperar 100ms
    workflow_jsonl = """{"operationUpdate": {"workflowId": "test", "operations": [
  {"id": "wait", "operation": {"Wait": {"duration": 100}}}
]}}
{"beginExecution": {"workflowId": "test", "root": "wait"}}"""
    
    print("\n1. Ejecutando workflow...")
    result = agent.execute_workflow_direct(workflow_jsonl, validate=True)
    print(f"   Estado: {result.get('status')}")
    print(f"   Execution ID: {result.get('execution_id')}")


async def example3_generate_workflow_with_llm():
    """Ejemplo 3: Generar workflow usando LLM"""
    print("\n" + "=" * 60)
    print("Ejemplo 3: Generar Workflow con LLM")
    print("=" * 60)
    
    agent = create_a2e_agent(
        a2e_server_url="http://localhost:8000",
        api_key="your-api-key-here"
    )
    
    # Pedir al agente que genere y ejecute un workflow
    print("\n1. Pidiendo al agente que genere un workflow...")
    response = await agent.run(
        "Genera y ejecuta un workflow que espere 1 segundo y luego haga una llamada GET a https://api.example.com/users"
    )
    print(f"   Respuesta: {response.content}")


async def example4_search_knowledge():
    """Ejemplo 4: Buscar conocimiento usando RAG"""
    print("\n" + "=" * 60)
    print("Ejemplo 4: Buscar Conocimiento (RAG)")
    print("=" * 60)
    
    agent = create_a2e_agent(
        a2e_server_url="http://localhost:8000",
        api_key="your-api-key-here"
    )
    
    print("\n1. Buscando conocimiento...")
    response = await agent.run("Busca información sobre cómo obtener usuarios de una API")
    print(f"   Respuesta: {response.content}")


async def example5_search_sql_queries():
    """Ejemplo 5: Buscar consultas SQL"""
    print("\n" + "=" * 60)
    print("Ejemplo 5: Buscar Consultas SQL")
    print("=" * 60)
    
    agent = create_a2e_agent(
        a2e_server_url="http://localhost:8000",
        api_key="your-api-key-here"
    )
    
    print("\n1. Buscando consultas SQL...")
    response = await agent.run("Busca consultas SQL para obtener usuarios activos")
    print(f"   Respuesta: {response.content}")


async def example6_workflow_validation():
    """Ejemplo 6: Validar workflow"""
    print("\n" + "=" * 60)
    print("Ejemplo 6: Validar Workflow")
    print("=" * 60)
    
    agent = create_a2e_agent(
        a2e_server_url="http://localhost:8000",
        api_key="your-api-key-here"
    )
    
    # Workflow con error potencial
    workflow_jsonl = """{"operationUpdate": {"workflowId": "test", "operations": [
  {"id": "fetch", "operation": {
    "ApiCall": {
      "method": "GET",
      "url": "https://api.example.com/users",
      "outputPath": "/workflow/users"
    }
  }}
]}}
{"beginExecution": {"workflowId": "test", "root": "fetch"}}"""
    
    print("\n1. Validando workflow...")
    response = await agent.run(f"Valida este workflow: {workflow_jsonl}")
    print(f"   Respuesta: {response.content}")


async def main():
    """Ejecuta todos los ejemplos"""
    print("\n" + "=" * 60)
    print("Ejemplos de Uso: A2E Agent con Google ADK")
    print("=" * 60)
    
    try:
        await example1_basic_usage()
        await example2_execute_workflow()
        await example3_generate_workflow_with_llm()
        await example4_search_knowledge()
        await example5_search_sql_queries()
        await example6_workflow_validation()
        
        print("\n" + "=" * 60)
        print("Todos los ejemplos completados")
        print("=" * 60)
    except Exception as e:
        print(f"\nError: {e}")
        import traceback
        traceback.print_exc()


if __name__ == "__main__":
    asyncio.run(main())

